packages/browser-wallet/src/popup/pages/AddTokens/ExternalAddTokens.tsx

```diff
@@ -0,0 +1,237 @@
+
+async function fetchTokenMetadata(
+    newIds: string[],
+    existingIds: string[],
+    client: JsonRpcClient,
+    network: NetworkConfiguration,
+    contractDetails: ContractDetails
+) {
+    const filteredIds =
+        // Remove duplicates
+        [...new Set(newIds)]
+            // Remove any ids that are not proper hex strings.
+            .filter((id) => isHex(id));
+
+    const metadataLinks = await getTokenUrl(client, filteredIds, contractDetails);
+    return Promise.all(
+        filteredIds.map(async (id, index) => {
+            const metadataLink = metadataLinks[index];
+            const metadata = await getTokenMetadata(metadataLink, network);
+            const status = existingIds?.some((cand) => cand === id) ? ChoiceStatus.existing : ChoiceStatus.chosen;
+            return { id, metadataLink, metadata, status };
+        })
+    );
+}
+
+interface Location {
+    state: {
+        payload: {
+            contractIndex: string;
+            contractSubindex: string;
+            tokenIds: string[];
+            url: string;
+        };
+    };
+}
+
+export default function SignMessage({ respond }: Props) {
+    const { state } = useLocation() as Location;
+    const { t } = useTranslation('account', { keyPrefix: 'tokens' });
+    const { withClose, onClose } = useContext(fullscreenPromptContext);
+    const { contractIndex, contractSubindex, tokenIds, url } = state.payload;
+    const addToast = useSetAtom(addToastAtom);
+    const client = useAtomValue(jsonRpcClientAtom);
+    const network = useAtomValue(networkConfigurationAtom);
+    const account = useAtomValue(selectedAccountAtom);
+    const [accountTokens, setAccountTokens] = useAtom(currentAccountTokensAtom);
+    const [addingTokens, setAddingTokens] = useState<TokenWithChoice[]>();
+
+    // If the window is closed, we avoid leaking information.
+    useEffect(() => onClose(() => respond([])), [onClose, respond]);
+
+    const onFinish = (newTokens: TokenIdAndMetadata[]) => {
+        setAccountTokens({ contractIndex, newTokens }).then(withClose(() => respond(newTokens.map(getId))));
+    };
+
+    const contractDetails = useAsyncMemo(
+        async () => {
+            const name = await fetchContractName(client, BigInt(contractIndex), BigInt(contractSubindex));
+            if (!name) {
+                throw new Error('Test');
+            }
+            return { index: BigInt(contractIndex), subindex: BigInt(contractSubindex), contractName: name };
+        },
+        (e) => addToast(e.message),
+        []
+    );
+
+    useEffect(() => {
+        if (!contractDetails || accountTokens.loading) {
+            return;
+        }
+        const existingIds = (accountTokens.value[contractIndex] || []).map((token) => token.id);
+        fetchTokenMetadata(tokenIds, existingIds, client, network, contractDetails).then((newTokens) =>
+            setAddingTokens(newTokens)
+        );
```

Comment:
I'll just filter them away, I already do that with weird ID's without saying. :thinking:

Reply:
I guess the dapp doesn' t know what's happening anyway, so should be fine not to state anything. It might be confusing if a dapp states "add x,y,z tokens to wallet" and you're only presented with a subset, if any?

Reply:
Ive added toasts.